modules/beefy/src/leaf.rs

```diff
@@ -0,0 +1,425 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! BEEFY MMR leaf verification verification.
+
+use crate::{
+	BridgedBeefyMmrHasher, BridgedBeefyMmrLeaf, BridgedBeefyMmrLeafUnpacked,
+	BridgedBeefyValidatorIdToMerkleLeaf, BridgedBeefyValidatorSet, BridgedBlockHash,
+	BridgedBlockNumber, Config, Error,
+};
+
+use bp_beefy::{
+	beefy_merkle_root, verify_mmr_leaf_proof, BeefyMmrHash, BeefyMmrProof, MmrDataOrHash,
+	MmrLeafVersion,
+};
+use codec::Decode;
+use frame_support::{ensure, traits::Get};
+use sp_runtime::{
+	traits::{Convert, One, Saturating},
+	RuntimeDebug,
+};
+use sp_std::marker::PhantomData;
+
+/// Artifacts of MMR leaf proof verification.
+#[derive(RuntimeDebug)]
+pub struct BeefyMmrLeafVerificationArtifacts<T: Config<I>, I: 'static> {
+	/// Block number and hash of the finalized block parent.
+	pub parent_number_and_hash: (BridgedBlockNumber<T, I>, BridgedBlockHash<T, I>),
+	/// Next validator set, if handoff is happening.
+	pub next_validator_set: Option<BridgedBeefyValidatorSet<T, I>>,
+	/// Parachain heads merkle root at the imported block.
+	pub parachain_heads: BeefyMmrHash,
+}
+
+/// Verify MMR proof of given leaf.
+///
+/// Returns new BEEFY validator set if it is enacted.
+pub fn verify_beefy_mmr_leaf<T: Config<I>, I: 'static>(
+	validators: &BridgedBeefyValidatorSet<T, I>,
+	at_header: BridgedBlockNumber<T, I>,
+	mmr_leaf: BridgedBeefyMmrLeafUnpacked<T, I>,
+	mmr_proof: BeefyMmrProof,
+	mmr_root: BeefyMmrHash,
+) -> Result<BeefyMmrLeafVerificationArtifacts<T, I>, Error<T, I>>
+where
+	BridgedBeefyMmrHasher<T, I>: 'static + Send + Sync,
+{
+	// decode raw MMR leaf
+	let raw_mmr_leaf = decode_raw_mmr_leaf::<T, I>(mmr_leaf.leaf())?;
+	ensure!(
+		raw_mmr_leaf.parent_number_and_hash.0 == at_header.saturating_sub(One::one()),
+		Error::<T, I>::InvalidParentNumberAndHash,
+	);
+
+	// TODO: is it the right condition? can id is increased by say +3?
```

Comment:
Actual question. My understanding of BEEFY is that:
1) commitment is signed by the current validator set, which must have id equal to `commitment.commitment.validator_set_id`;
2) if `mmr_leaf.beefy_next_authority_set.id == commitment.commitment.validator_set_id + 1`, then it is the regular header and its direct descendant will be signed by the same set `commitment.commitment.validator_set_id`;
3) when we see `mmr_leaf.beefy_next_authority_set.id == commitment.commitment.validator_set_id + 2`, then it is the 'handoff' header, which must be signed by current validator set. But its direct descendant will be signed by the next set `commitment.commitment.validator_set_id + 1`.

Is that correct? I'll be able to find it later, during relay development, but people who have more BEEFY knowledge may answer earlier.

Follow-up question is whether we may see e.g. `mmr_leaf.beefy_next_authority_set.id == commitment.commitment.validator_set_id + 3` here? Is it valid to skip set id in BEEFY? Like jumping from set `10` to `15`?

Reply:
> you are referring to mmr_leaf (which represents/identifies a particular block basically) and commitment - is the commitment here a vote for block number matching mmr_leaf, or the full commitment (super-majority set of votes finalizing the block)?

That's the full commitment (`SignedCommitment`). I mean - we don't have (and don't need) an access to individual votes here, especially in votes that haven't made it in the full commitment. So this light client only works with `SignedCommitment`s

Reply:
> not valid to skip set ids, validator set changes each session (even if validators and keys stay the same) and set id increments monotonically.

Good to know, thank you!